pex/cli/commands/lock.py

```diff
@@ -1203,9 +1229,251 @@ def add_warning(
+            }
+            retain = set()
+            to_resolve = deque(root_requirements)
+            while to_resolve:
+                req = to_resolve.popleft()
+                if req.project_name in retain:
+                    continue
+                retain.add(req.project_name)
+                dep = available.get(req.project_name)
+                if not dep:
+                    return Error(
+                        "There is no lock entry for {project} in {lock_file} to satisfy the "
+                        "{transitive}'{req}' requirement.".format(
+                            project=req.project_name,
+                            lock_file=lockfile_path,
+                            transitive="" if req in root_requirements else "transitive ",
+                            req=req,
+                        )
+                    )
+                elif dep:
+                    pnav, locked_req = dep
+                    if pnav not in req:
+                        production_assert(
+                            req in root_requirements,
+                            "Transitive requirements in a lock should always match existing lock "
+                            "entries. Found {project} {version} in {lock_file}, which does not "
+                            "satisfy transitive requirement '{req}' found in the same lock.".format(
+                                project=pnav.project_name,
+                                version=pnav.version,
+                                lock_file=lockfile_path,
+                                req=req,
+                            ),
+                        )
+                        return Error(
+                            "The locked version of {project} in {lock_file} is {version} which "
+                            "does not satisfy the '{req}' requirement.".format(
+                                project=pnav.project_name,
+                                lock_file=lockfile_path,
+                                version=pnav.version,
+                                req=req,
+                            )
+                        )
+                    elif (
+                        req.project_name in constraint_by_project_name
+                        and pnav not in constraint_by_project_name[req.project_name]
+                    ):
+                        return Error(
+                            "The locked version of {project} in {lock_file} is {version} which "
+                            "does not satisfy the '{constraint}' constraint.".format(
+                                project=pnav.project_name,
+                                lock_file=lockfile_path,
+                                version=pnav.version,
+                                constraint=constraint_by_project_name[req.project_name],
+                            )
+                        )
+                    to_resolve.extend(requires_dist.filter_dependencies(req, locked_req))
```

Comment:
This was the only trick bit. To correctly subset you need to consider extras but no other marker expressions. Luckily, the `--elide-unused-requires-dists` feature stabilized in https://github.com/pex-tool/pex/releases/tag/v2.25.2 included this machinery. The new integration tests cover this case.

Reply:
You also need to consider the markers faked for `--style universal` locks, namely "os_name", "platform_system" & "sys_platform" (here: https://github.com/pex-tool/pex/blob/16932ad0412ed3019612c5ef604362554169b7ed/pex/resolve/locker.py#L568-L583) and "python_version" & "python_full_version" (here: https://github.com/pex-tool/pex/blob/16932ad0412ed3019612c5ef604362554169b7ed/pex/pip/foreign_platform/__init__.py#L181-L206). #2684 fixes this pretty huge oversight.